### DISCUSSION TITLE
Oversized github repository

### DISCUSSION THREAD

**Discussion by michaelchin** [GPlates/gplately#288] [Ideas]
The size of this repository is 914M as of 13 Jun 2023. Write down this problem lest I forgot.

`du -sh -- * .[^.]* | sort -h`

<img width="600" alt="Screenshot 2023-06-13 at 6 05 30 pm" src="https://github.com/GPlates/gplately/assets/2688316/793bac1a-aaa6-42ad-bacb-5395425a1dc0">

Ignore history: 

`git clone --depth 1 --branch master https://github.com/GPlates/gplately.git`

**Comment by brmather** [comment]
All of this is taken up by graphics and animations in the notebooks. I would like to clear these from all notebooks in the repo and flush them from the git history as well.

I've looked at a few solutions in the past but none were fantastic. Running `jupyter nbconvert` in a git hook might be one way. Alternatively we could use jupytext to properly version control notebooks as markdown files (or regular python files).

**Comment by michaelchin** [comment]
Just a note to myself.

use BFG to purge history https://rtyley.github.io/bfg-repo-cleaner/#download

**Comment by michaelchin** [comment]
Put this on ice until it causes real problems or too painful to bear.

**Comment by brmather** [comment]
> use BFG to purge history https://rtyley.github.io/bfg-repo-cleaner/#download

Most of the space is taken up by Jupyter notebook outputs (images, gifs, etc.) Does BFG help with this?

I've found a couple more resources which may help with Jupyter notebooks:

https://zhauniarovich.com/post/2020/2020-10-clearing-jupyter-output-p3/
https://www.scivision.dev/git-jupyter-strip-output/

**Comment by michaelchin** [comment]
> > use BFG to purge history https://rtyley.github.io/bfg-repo-cleaner/#download
> 
> Most of the space is taken up by Jupyter notebook outputs (images, gifs, etc.) Does BFG help with this?
> 
> I've found a couple more resources which may help with Jupyter notebooks:
> 
> https://zhauniarovich.com/post/2020/2020-10-clearing-jupyter-output-p3/ https://www.scivision.dev/git-jupyter-strip-output/

BFG just purges history. The history may take up some  space. I think we can try your findings first. This issue is not too painful for now.

Anyway, people can always ignore history by 

`git clone --depth 1 --branch master https://github.com/GPlates/gplately.git`

See the screenshot below. The history is not a big problem anymore.

`du -sh -- * .[^.]* | sort -h`

<img width="478" alt="Screenshot 2024-07-11 at 1 54 10 PM" src="https://github.com/GPlates/gplately/assets/2688316/ad4f4087-0497-467a-8888-ba5e06e4e075">

**Reply by jcannon-gplates** [reply]
Just commenting because you raised this issue again recently by email (since we're probably going to be adding a lot of notebooks as part of the documentation overhaul)...

From your screenshot it looks like the `Notebooks` directory is taking most of the space (compared to the `.git` history).

What Ben suggested [here](https://github.com/GPlates/gplately/discussions/288#discussioncomment-10920855) sounds good to me. I guess you are already using `nbconvert` to generate the documentation (ie, convert notebooks to html). Perhaps the documentation is the only place where output cells need to be visible? Does that mean the notebooks themselves would have no graphics or animations (that is, until the user actually runs them to generate the output)?

**Reply by jcannon-gplates** [reply]
This is what I get on a newly cloned repo (master branch at 7b267fd)...

![image](https://github.com/user-attachments/assets/ac7924c2-cb6f-4f90-a2a1-f8adc2ed770c)

...looks like the `.git` folder has increased dramatically from Michael's 80 MB to my 875 MB. And pretty much all of that is coming from `.git/objects/`.

So it looks like the history is becoming the biggest problem.

While the `Notebooks/` sub-directory is reasonably big...

![image](https://github.com/user-attachments/assets/25d029e5-31ee-4c55-a5b6-d5e9e181bb2b)

...the actual notebooks themselves only take up about 50 MB (of that 235 MB `Notebooks/` sub-directory).

But committing notebooks with updated plots/animations is really accumulating in the git history.

For example, I just updated all the notebooks (including their plots/animations) in #337 and the size of `.git/objects` increased by 40 MB. If I were to do that again it would increase another 40 MB.

And, as a small test on the newly cloned repo, committing a notebook with a small change (just some code change) added less than 1 MB. But when I updated its plots/animations and committed, then it added about 5 MB. I guess that's pretty obvious, but just confirms that git is pretty good at delta-compressing small text changes in a notebook when all the binary bits (plots/animations) are the same.

So, in the short term, we could update only the code in the notebooks without updating plots/animations (if they don't change much), and save space that way. But probably looks like a longer term solution (like Ben and Michael have suggested) will be needed fairly soon.

**Reply by jcannon-gplates** [reply]
I just realized that the 80 MB of `.git` history in Michael's case was on a cloned repo of depth 1.

Still, it looks like the history size has increased from 700 MB in mid-2023 (top post above) to over 900 MB now.

**Reply by michaelchin** [reply]
> I just realized that the 80 MB of `.git` history in Michael's case was on a cloned repo of depth 1.
> 
> Still, it looks like the history size has increased from 700 MB in mid-2023 (top post above) to over 900 MB now.

Thanks John. It seems the the history size only increased ~200MB over two years. In fact, it was better than I thought. Yes, the outputs(images/animations) in the notebooks are the problem. I think notebooks using base64 encoding to save those images make the situation even worse.

**Reply by jcannon-gplates** [reply]
Yeah, that's not too bad. Maybe that's because we haven't really modified the notebooks much over the last couple of years. As we start adding more and modifying it might accelerate a bit. We'll find out soon enough though.

**Comment by michaelchin** [comment]
convert to discussion until we decide some actionable plans.
